src/providers/wfs/qgsbackgroundcachedfeatureiterator.cpp

```diff
@@ -766,7 +766,7 @@ bool QgsBackgroundCachedFeatureIterator::fetchFeature( QgsFeature &f )
       mWaitCond.wait( &mMutex, timeout );
       if ( requestMadeFromMainThread && mProcessEvents )
       {
-        QgsApplication::instance()->processEvents();
+        QApplication::instance()->processEvents();
```

Comment:
Totally unrelated with this PR I know, I'm not asking to fix it now, I just want to start a discussion. 

Is there a way we can avoid processing events here? It is causing a lot of problems in the server context when we want to wrap it in an external QTcpServer (like in the standalone server). 

The same issue exists in WMS provider.

Reply:
This is using the same solution as in QgsBlockingNetworkRequest, that is to be able to deal with authentication :

commit daf78c64e9e5342a7f289366a070db46dbcacc36
Author: Even Rouault <even.rouault@spatialys.com>
Date:   Tue Sep 22 15:22:51 2020 +0200

    [WFS/OAPIF providers] Fix freeze when feature requests issued from main thread and authentication manager triggers (fixes #37224)

Reply:
In the server context we don't need to interact with the users to enter the credentials, I was wondering if there is a way to bypass this code for the the server.

Reply:
So this particular loop will only be called if the QgsNetworkAccessManager "authRequestOccurred" , "proxyAuthenticationRequired" or "sslErrorsOccurred" signals are called (because these all trigger dialogs for user interaction on desktop). 

I think the fixes would need to sit within QgsNetworkAccessManager, and e.g hook into the QgsSslErrorHandler to indicate whether its required to run the event loop. Server would then implement a QgsSslErrorHandler which denies this, while app would allow it. (And similar for the proxy/auth request handling).

Reply:
(In any case, the fix resides outside of the WFS client)

Reply:
Ok, good to know, the server doesn't set and use any QgsSslErrorHandler so that's not an issue. Also it uses the default QgsCredentialsNone. I think the only problem left to solve is in the WMS provider eventLoop.